Feature/app shut down

tmengesh opened the pull request:
## We welcome contributions from the public. However there are some guidelines to follow:

1. We cannot accept new language versions and translations from community, this is because:
	- Some expressions used within the app are strictly coupled with Finnish law and thus must meet it's requirements
	- There's also a lot of domain specific terminology involved which must be handled by THL's board of experts
	- Small typo fixes can be processed manually but better way to report those is to raise an issue. We can then manually credit changes to issue's original author
2. Prior to committing to work, **create an issue** describing a problem, bug or enhancement you would to like to work on
3. **Wait for discussion** on the issue and how it should be solved
4. **Wait for main contributors** of the repository to handle the issue and clear it for implementation
5. Embrace the feedback and be patient. We are working as fast as we can to improve Koronavilkku and community's help is much appreciated

Please **always** add reference to related issue. You can bring up drop-down of suggested issues by typing #.

This pull request resolves # .

Also, remember to update CHANGELOG.md before merging your changes.

tmengesh committed:
Create EndOfLifeViewController + StatisticsCard

tmengesh committed:
EndOfLife localizable strings

tmengesh committed:
Adding EndOfLifeViewController + StatisticsCard to the project

tmengesh committed:
Fetching and assigning EndOfLifeStatistic to localStore

tmengesh committed:
Present EndOfLifeViewController and do cleanup if endOfLife reached.

tmengesh committed:
Update EndOfLifeViewController.swift

tmengesh committed:
Update BackendRestApiTest.swift

tmengesh committed:
Refactor EndOfLife check and cleanup.

tmengesh committed:
Refactor EndOfLife check

tmengesh committed:
Handling endOfLife case for background tasks.

spheroid commented on Koronavilkku/Koronavilkku/Model/ExposureConfiguration.swift:
```diff
@@ -18,6 +18,10 @@ struct ExposureConfiguration: Codable {
+    
+    let endOfLifeReached: Bool
+    let endOfLifeStatistics: [EndOfLifeStatistic]
+   // let endOfLifeStatistics: [String : [String : String]]
```
No need to have this here 🙂

spheroid commented on Koronavilkku/Koronavilkku/SceneDelegate.swift:
```diff
@@ -67,10 +72,23 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate, UNUserNotificationCente
     }
 
     func sceneDidBecomeActive(_ scene: UIScene) {
+        guard LocalStore.shared.endOfLifeStatisticsData.isEmpty else {
```
You should probably use the isEndOfLife() here as well.

spheroid commented on Koronavilkku/Koronavilkku/ViewController/EndOfLifeViewController.swift:
```diff
@@ -0,0 +1,85 @@
+    var titleText = Translation.EndOfLifeTitle.localized
+    var textLabelText = Translation.EndOfLifeMessage.localized
+    
+    var configurationTask: AnyCancellable?
```
This is not used

tmengesh replied:
I've removed configurationTask, however, titleText and textLabelText are used while setting up title and description.

spheroid commented on Koronavilkku/Koronavilkku/UI/Components/StatisticsCard.swift:
```diff
@@ -0,0 +1,74 @@
+        
+        container.clipsToBounds = true
+        container.layer.cornerRadius = cornerRadius
+        titleView.isAccessibilityElement = false
+        bodyView.isAccessibilityElement = false
+        
+        rightPadding.accessibilityTraits = .none
+        
+        self.accessibilityTraits = .button
+        self.isAccessibilityElement = true
+        self.accessibilityLabel = titleView.text
+        self.accessibilityValue = bodyView.text
```
As this is not a button, you could probably stick to the default behaviour

spheroid commented on Koronavilkku/Koronavilkku/Repository/ExposureRepository.swift:
```diff
@@ -105,9 +106,25 @@ struct ExposureRepositoryImpl : ExposureRepository {
-        return backend.getConfiguration()
+        return backend.getConfiguration().map { config in
+            if config.endOfLifeReached {
+                LocalStore.shared.endOfLifeStatisticsData = config.endOfLifeStatistics
+                notificationService.updateBadgeNumber(nil)
+                setStatus(enabled: false)
+                LocalStore.shared.exposures.removeAll()
+                LocalStore.shared.countExposureNotifications.removeAll()
+                LocalStore.shared.daysExposureNotifications.removeAll()
```
These are not run in main thread, perhaps add a `receive(on: RunLoop.main)` modifier to the chain before the `.map()`

tmengesh replied:
👍

spheroid commented on Koronavilkku/Koronavilkku/ViewController/EndOfLifeViewController.swift:
The scrollview is not bound to the bottom, you need to

```
        logoImageView.snp.makeConstraints { make in
            make.bottom.equalToSuperview()
        }
```

tmengesh replied:
Done :)

spheroid replied:
I can't see these changes yet here

tmengesh replied:
Now?
https://github.com/THLfi/koronavilkku-ios/pull/62/commits/393997ad9fd82673e4c9ee5767d7be4bbd431403

spheroid commented on Koronavilkku/Koronavilkku/ViewController/EndOfLifeViewController.swift:
```diff
@@ -0,0 +1,85 @@
+        
+        //Setup title
+        titleLabel.text = titleText
+        titleLabel.font = .heading1
```
Should be .heading2

spheroid commented on Koronavilkku/Koronavilkku/UI/Components/StatisticsCard.swift:
```diff
@@ -0,0 +1,74 @@
+import UIKit
+
+final class StatisticsCard: WideRowElement {
```
Perhaps plain CardElement would suit better here?

tmengesh replied:
<img width="267" alt="Screenshot 2022-03-23 at 15 56 18" src="https://user-images.githubusercontent.com/17072260/159716094-591d5c17-2a3d-4830-a45f-f9ccee87710b.png">

This is how the view looks with CardElement

tmengesh replied:
I think it suits better.

tmengesh committed:
using isEndOfLife()'s value instead of checking if endOfLifeStatisticsData is empty

tmengesh committed:
Changed WideRowElement with plain CardElement

tmengesh committed:
Update ExposureRepository.swift

tmengesh committed:
Added bottom constraint to logo

spheroid commented on Koronavilkku/Koronavilkku/SceneDelegate.swift:
```diff
@@ -67,10 +72,23 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate, UNUserNotificationCente
         
+       configurationTask = Environment.default.exposureRepository.getConfiguration().sink { _ in } receiveValue: { _ in
+           if (Environment.default.exposureRepository.isEndOfLife()) {
+                DispatchQueue.main.async {
```
Oh, this could've also be replaced with `receive(on:)` before the sink

tmengesh replied:
Done

spheroid replied:
…and therefore you no longer need the DispatchQueue.main.async 😄

tmengesh committed:
Cleanup + changing titleLable font to heading2

tmengesh committed:
Update ExposureConfiguration.swift

tmengesh committed:
Update SceneDelegate.swift

tmengesh committed:
Update EndOfLifeViewController.swift

tmengesh committed:
Refactored Locale for StatisticsCard

tmengesh committed:
receive(on: RunLoop.main) instead of DispatchQueue.main.async

spheroid reviewed (commented):
Looks good. If nothing comes up in the device tests, this is ready to 🚢 .

tmengesh committed:
Update EndOfLifeViewController.swift

tmengesh committed:
Enabling Dynamic Type change to all UILabels within the app.

Moved Accessibility (UILabel responding to Dynamic type change) from RootViewController to SceneDelegate to support a dynamic change to all views within the app.

Commit:
Move EFGS country list refresh to SceneDelegate

Commit:
Use pre-existing Localized

Commit:
Detect the end of life event and update the UI

Commit:
Don't accidentally request turning on the EN API

If the user installs the app after end of life event, the API is not enabled and yet "turning it off" will cause a popup asking whether the user wants to turn it on.

Commit:
Clean up

Commit:
Update CHANGELOG.md

Commit:
Treat statistics item as an accessibility element

spheroid commented:
Noticed some things during the on-device testing and took the liberty to fix them:

- The app might also detect the end of life during manual exposure check, which would perform the EOL actions but wouldn't cause the UI to change. Added a new publisher that emits when the EOL is detected and changed the SceneDelegate to use it to update the UI.

- Noticed that the configuration was fetched twice during the app launch due to EFGS sync. Combined the EOL and EFGS refresh logic in the same configuration fetch operation.

- VoiceOver treated the statistics element as two separate text labels, which they semantically aren't.

- If the user installed the app after EOL, the turning API off event would actually ask permissions to use the API. Added an authorisation status check before attempting to turn it off.

Commit:
Update CHANGELOG.md
